SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NetDriver](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			ChildActor->ForceNetUpdate();
+			Channel->MigrationVirtualWorkerId.Emplace(SpatialConstants::INVALID_VIRTUAL_WORKER_ID);
+			Channel->MigrationInitiatedFrame = NetDriver->ReplicationFrame;
+		}
+	}));
+}
+
+void USpatialActorChannel::SetActorHierarchyMigrationDestination(USpatialNetDriver* NetDriver, VirtualWorkerId NewAuthVirtualWorkerId, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NewAuthVirtualWorkerId] (AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if(ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId = NewAuthVirtualWorkerId;
+		}
+	}));
+}
+
+void USpatialActorChannel::DropActorHierarchyMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId.Reset();
+		}
+	}));
+}
+
+void USpatialActorChannel::HandleActorLoadBalancing()
+{
+	// if the field is set it means we should try to migrate authority to another worker
+	if (MigrationVirtualWorkerId.IsSet())
+	{
+		AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+
+		// If we were locked in between replication frames, drop migration attempt.
+		if (NetDriver->LockingPolicy->IsLocked(Actor))
+		{
+			DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+		}
+		// Are we in a different frame from the one we determined we should try to migrate ?
+		else if (MigrationInitiatedFrame != NetDriver->ReplicationFrame)
+		{
+			// An invalid worker id means we encounter the first actor in the hierarchy, which will determine where all the other actors will migrate.
+			if (MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
 			{
-				OutTimestamp = Channel->AuthorityReceivedTimestamp;
+				const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*HierarchyRoot);
+				if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+				{
+					UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+					DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+				}
+				else if (NewAuthVirtualWorkerId == NetDriver->LoadBalanceStrategy->GetLocalVirtualWorkerId())
+				{
+					DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+				}
+				else
+				{
+					SetActorHierarchyMigrationDestination(NetDriver, NewAuthVirtualWorkerId, HierarchyRoot);
+				}
+			}
+
+			if (const VirtualWorkerId NewAuthVirtualWorkerId = MigrationVirtualWorkerId.GetValue())
```

Comment:
What's the reason for declaring `NewAuthVirtualWorkerId`? It's not used in the block. Also, I'm assuming this is supposed to check for whether `MigrationVirtualWorkerId` is valid, in which case I'd expect it to be compared against `INVALID_VIRTUAL_WORKER_ID`.